cupy/cutensor.pyx

```diff
@@ -2,46 +2,72 @@ import numpy
 import warnings
 
 import cupy
-from cupy_backends.cuda.api import runtime
+
+from libc.stdint cimport intptr_t, uint32_t, uint64_t
+from cupy.core._carray cimport shape_t
+from cupy.core.core cimport ndarray
+from cupy.core cimport internal
+from cupy_backends.cuda.libs.cutensor cimport Handle
+from cupy_backends.cuda.libs.cutensor cimport TensorDescriptor
+from cupy_backends.cuda.libs.cutensor cimport ContractionDescriptor
+from cupy_backends.cuda.libs.cutensor cimport ContractionFind
+from cupy_backends.cuda.libs.cutensor cimport ContractionPlan
+
+from cupy.core cimport core
+from cupy.core cimport _reduction
+from cupy.cuda cimport device
+from cupy_backends.cuda.api cimport runtime
 from cupy_backends.cuda.libs cimport cutensor
-from cupy.cuda import device
-from cupy.core import _reduction
 
-_handles = {}
-_tensor_descriptors = {}
-_contraction_descriptors = {}
-_contraction_finds = {}
-_contraction_plans = {}
 
+cdef dict _handles = {}
+cdef dict _tensor_descriptors = {}
+cdef dict _contraction_descriptors = {}
+cdef dict _contraction_finds = {}
+cdef dict _contraction_plans = {}
+
+
+cdef class Mode(object):
 
-class Mode(object):
+    cdef:
+        object _array
+        readonly int ndim
+        readonly intptr_t data
 
     def __init__(self, mode):
-        self.mode = numpy.array(mode, dtype=numpy.int32)
-        assert self.mode.ndim == 1
+        self._array = numpy.array(mode, dtype=numpy.int32)
+        assert self._array.ndim == 1
+        self.ndim = self._array.size
+        self.data = self._array.ctypes.data
+
+    def __repr__(self):
+        return 'mode([' + ', '.join(self._array) + '])'
 
-    @property
-    def ndim(self):
-        return self.mode.size
 
-    @property
-    def data(self):
-        return self.mode.ctypes.data
+cdef class _Scalar(object):
+
+    cdef:
+        object _array
+        readonly intptr_t ptr
+
+    def __init__(self, value, dtype):
+        self._array = numpy.asarray(value, dtype=dtype)
+        self.ptr = self._array.ctypes.data
 
     def __repr__(self):
-        return 'mode([' + ', '.join(self.mode) + '])'
+        return self._array.item()
 
 
-def get_handle():
-    dev = device.get_device_id()
+cpdef Handle get_handle():
+    cdef int dev = device.get_device_id()
     if dev not in _handles:
-        handle = cutensor.Handle()
+        handle = Handle()
         cutensor.init(handle)
         _handles[dev] = handle
     return _handles[dev]
 
 
-def get_cuda_dtype(numpy_dtype):
+cpdef int get_cuda_dtype(numpy_dtype):
```

Comment:
I wanted to add `except -1` here, but I think this function will be removed in favor of the change in #3719, so it's fine to not touch it.

Reply:
Fixed in 1b16b96. Thanks!

Reply:
I think you can cimport and reuse this function:
https://github.com/cupy/cupy/blob/c3afff0e34fab6325c3db77da442ca5d24882b1c/cupy/core/_dtype.pxd#L3

Reply:
`to_cuda_dtype` works only for `numpy.dtype` object. It does not work if `compute_dtype` of str type is given. (e.g., `cutensor.reduction(..., compute_dtype='d')`)

Reply:
1. I skimmed over and don't think you have passed str types to `get_cuda_dtype`
2. I still think it's best to reuse the same function whenever possible. If a str type argument is needed, how about revising `to_cuda_dtype` in `_dtype.pyx` this way?
```python
    if dtype in ('f', numpy.float32):
        return runtime.CUDA_R_32F
    ...
```